core/src/main/java/org/infinispan/persistence/support/SegmentedAdvancedLoadWriteStoreAdapter.java

```diff
@@ -0,0 +1,205 @@
+import org.infinispan.persistence.spi.ExternalStore;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.MarshallableEntry;
+import org.infinispan.persistence.spi.SegmentedAdvancedLoadWriteStore;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+
+public class SegmentedAdvancedLoadWriteStoreAdapter<K, V> implements SegmentedAdvancedLoadWriteStore<K, V> {
+   private final CacheLoader<K, V> cacheLoader;
+   private final CacheWriter<K, V> cacheWriter;
+
+   private KeyPartitioner keyPartitioner;
+
+   public SegmentedAdvancedLoadWriteStoreAdapter(CacheLoader<K, V> cacheLoader) {
+      this.cacheLoader = Objects.requireNonNull(cacheLoader);
+      this.cacheWriter = cacheLoader instanceof CacheWriter ? (CacheWriter) cacheLoader : null;
+   }
+
+   public SegmentedAdvancedLoadWriteStoreAdapter(CacheWriter<K, V> cacheWriter) {
+      this.cacheWriter = Objects.requireNonNull(cacheWriter);
+      this.cacheLoader = cacheWriter instanceof CacheLoader ? (CacheLoader) cacheWriter : null;
+   }
+
+   @Override
+   public void init(InitializationContext ctx) {
+      keyPartitioner = ctx.getKeyPartitioner();
+      if (cacheLoader != null) {
+         cacheLoader.init(ctx);
+      } else {
+         cacheWriter.init(ctx);
+      }
+   }
+
+   @Override
+   public void start() {
+      if (cacheLoader != null) {
+         cacheLoader.start();
+      } else {
+         cacheWriter.start();
+      }
+   }
+
+   @Override
+   public void stop() {
+      if (cacheLoader != null) {
+         cacheLoader.start();
+      } else {
+         cacheWriter.start();
+      }
+   }
+
+   @Override
+   public void destroy() {
+      // ExternalStore is both loader and writer so can just check one
+      if (cacheLoader instanceof ExternalStore) {
+         ((ExternalStore<K, V>) cacheLoader).destroy();
+      } else {
+         stop();
+      }
+   }
+
+   @Override
+   public boolean isAvailable() {
+      if (cacheLoader != null) {
+         return cacheLoader.isAvailable();
+      } else {
+         return cacheWriter.isAvailable();
+      }
+   }
+
+   @Override
+   public MarshallableEntry<K, V> loadEntry(Object key) {
+      return cacheLoader.loadEntry(key);
+   }
+
+   @Override
+   public MarshallableEntry<K, V> get(int segment, Object key) {
+      return loadEntry(key);
+   }
+
+   @Override
+   public boolean contains(Object key) {
+      return cacheLoader.contains(key);
+   }
+
+   @Override
+   public boolean contains(int segment, Object key) {
+      return contains(key);
+   }
+
+   @Override
+   public void write(MarshallableEntry<? extends K, ? extends V> entry) {
+      cacheWriter.write(entry);
+   }
+
+   @Override
+   public void write(int segment, MarshallableEntry<? extends K, ? extends V> entry) {
+      write(entry);
+   }
+
+   @Override
+   public boolean delete(Object key) {
+      return false;
```

Comment:
why return `false`? 
shouldn't the "write" methods check if `cacheWriter` is `null`? same for "read" methods.